Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,321 @@
+    /// This function takes  a `reservedCapacity` parameter which is the number of bytes we already reserved for writing this integer.
+    /// If you use write more or less than this many bytes, we wll need to move bytes around to make that possible.
+    /// Therefore, you may decide to encode differently to use more bytes than you actually would need, if it is possible for you to do so.
+    /// That way, you waste the reserved bytes, but improve writing performance.
+    func writeIntegerWithReservedCapacity(
+        _ integer: Int,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int
+}
+
+extension NIOBinaryIntegerEncodingStrategy {
+    @inlinable
+    public var reservedCapacityForInteger: Int { 1 }
+
+    @inlinable
+    public func writeIntegerWithReservedCapacity<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int {
+        self.writeInteger(integer, to: &buffer)
+    }
+}
+
+extension ByteBuffer {
+    /// Read a binary encoded integer, moving the `readerIndex` appropriately.
+    /// If there are not enough bytes, nil is returned.
+    @inlinable
+    public mutating func readEncodedInteger<Strategy: NIOBinaryIntegerEncodingStrategy>(_ strategy: Strategy) -> Int? {
+        strategy.readInteger(as: Int.self, from: &self)
+    }
+
+    /// Write a binary encoded integer.
+    ///
+    /// - Returns: The number of bytes written.
+    @discardableResult
+    @inlinable
+    public mutating func writeEncodedInteger<
+        Integer: FixedWidthInteger,
+        Strategy: NIOBinaryIntegerEncodingStrategy
+    >(
+        _ value: Integer,
+        strategy: Strategy
+    ) -> Int {
+        strategy.writeInteger(value, to: &self)
+    }
+
+    /// Prefixes a message written by `writeMessage` with the number of bytes written using `strategy`.
+    ///
+    /// - Note: This function works by reserving the number of bytes suggested by `strategy` before the data.
+    /// It then writes the data, and then goes back to write the length.
+    /// If the reserved capacity turns out to be too little or too much, then the data will be shifted.
+    /// Therefore, this function is most performant if the strategy is able to use the same number of bytes that it reserved.
+    ///
+    /// - Parameters:
+    ///     - strategy: The strategy to use for encoding the length.
+    ///     - writeMessage: A closure that takes a buffer, writes a message to it and returns the number of bytes written.
+    /// - Returns: Number of total bytes written. This is the length of the written message + the number of bytes used to write the length before it.
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefixed<Strategy: NIOBinaryIntegerEncodingStrategy>(
+        strategy: Strategy,
+        writeMessage: (inout ByteBuffer) throws -> Int
```

Comment:
nit: name closure parameters (per the [API design guidelines](https://www.swift.org/documentation/api-design-guidelines/#special-instructions)) as they act as documentation

```suggestion
        writeMessage: (_ buffer: inout ByteBuffer) throws -> Int
```